### PR TITLE
PMM-7385: MySQL restore job

### PR DESCRIPTION
**Pull request by oter**
https://jira.percona.com/browse/PMM-7385
API PR: https://github.com/percona/pmm/pull/717

### CODE REVIEW

File: jobs/mysql_restore_job.go [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+// BackupLocationConfig groups all backup locations configs.
+type BackupLocationConfig struct {
+	S3Config *S3LocationConfig
+}
+
+func NewMySQLRestoreJob(id string, timeout time.Duration, name string, locationConfig BackupLocationConfig) *MySQLRestoreJob {
+	return &MySQLRestoreJob{
+		id:       id,
+		timeout:  timeout,
+		l:        logrus.WithFields(logrus.Fields{"id": id, "type": "mysql_backup"}),
```
Comment (Dasio) [inline jobs/mysql_restore_job.go:74]:
```suggestion
		l:        logrus.WithFields(logrus.Fields{"id": id, "type": "mysql_restore"}),
```

File: jobs/mysql_restore_job.go [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+// stdout and stderr could be returned even if rerr is not nil
+func restoreMySQLFromS3(
+	ctx context.Context,
+	backupName string,
+	config *BackupLocationConfig,
+	targetDirectory string,
+) (stdout, stderr *bytes.Buffer, rerr error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer func() {
```
Comment (Dasio) [inline jobs/mysql_restore_job.go:124]: why not just defer cancel()?

**Review by Dasio — APPROVED** [review]

File: jobs/mysql_restore_job.go [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+	qpressBin           = "qpress"
+	mySQLServiceName    = "mysql"
+	mySQLUserName       = "mysql"
+	mySQLGroupName      = "mysql"
+	mySQLDirectory      = "/var/lib/mysql"
+	stopTimeout         = 5 * time.Second
+	activeCheckInterval = time.Second
+)
+
+type MySQLRestoreJob struct {
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply jobs/mysql_restore_job.go:48]: Yep, it's required)

File: jobs/mysql_restore_job.go [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+	backupName string,
+	config *BackupLocationConfig,
+	targetDirectory string,
+) (stdout, stderr *bytes.Buffer, rerr error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer func() {
+		cancel()
+	}()
+
+	xbcloudCmd := exec.CommandContext(ctx, xbcloudBin,
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply jobs/mysql_restore_job.go:128]: Mute that warning please

File: jobs/mysql_restore_job.go [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+	return nil
+}
+
+// stdout and stderr could be returned even if rerr is not nil
+func restoreMySQLFromS3(
+	ctx context.Context,
+	backupName string,
+	config *BackupLocationConfig,
+	targetDirectory string,
+) (stdout, stderr *bytes.Buffer, rerr error) {
```
Comment (artemgavrilov) [inline jobs/mysql_restore_job.go:122]: Can we eliminate named function results? Typically it's harder to understand what we return.

Comment (oter) [reply jobs/mysql_restore_job.go:122]: I'm not using stdout and stderr variables, it's just for indicating what returned in what order.

File: jobs/mysql_restore_job.go (hunk 2 above)
Comment (oter) [reply jobs/mysql_restore_job.go:124]: how this happened 😄

File: jobs/mysql_restore_job.go (hunk 5 above)
Comment (artemgavrilov) [reply jobs/mysql_restore_job.go:122]: And `rerr` ?

File: jobs/mysql_restore_job.go [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+
+	if _, err := exec.LookPath(qpressBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", qpressBin)
+	}
+
+	return nil
+}
+
+// stdout and stderr could be returned even if rerr is not nil
+func restoreMySQLFromS3(
```
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go:117]: split into two functions

File: jobs/mysql_restore_job.go (hunk 5 above)
Comment (artemgavrilov) [reply jobs/mysql_restore_job.go]: >  it's just for indicating what returned in what order.
Better to use comments for that

File: jobs/mysql_restore_job.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go:50]: done

File: jobs/mysql_restore_job.go [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+	BucketName   string
+	BucketRegion string
+}
+
+// BackupLocationConfig groups all backup locations configs.
+type BackupLocationConfig struct {
+	S3Config *S3LocationConfig
+}
+
+func NewMySQLRestoreJob(id string, timeout time.Duration, name string, locationConfig BackupLocationConfig) *MySQLRestoreJob {
```
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go:72]: done

File: jobs/mysql_restore_job.go [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+func (j *MySQLRestoreJob) ID() string {
+	return j.id
+}
+
+// Type returns job type.
+func (j *MySQLRestoreJob) Type() string {
+	return "mysql_restore"
+}
+
+// Timeouts returns job timeout.
```
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go:93]: done

File: jobs/mysql_restore_job.go [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+
+	var stderrBuffer bytes.Buffer
+	xbcloudCmd.Stderr = &stderrBuffer
+
+	xbcloudStdout, err := xbcloudCmd.StdoutPipe()
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to get xbcloud stdout pipe")
+	}
+
+	xbstreamCmd := exec.CommandContext(ctx, xbstreamBin,
```
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go:146]: muted

File: jobs/mysql_restore_job.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go:127]: muted

File: jobs/mysql_restore_job.go [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+		return true, nil
+	case os.IsNotExist(err):
+		return false, nil
+	default:
+		return false, errors.WithStack(err)
+	}
+}
+
+func restoreBackup(backupDirectory, mySQLDirectory string) error {
+	if _, err := exec.Command(xtrabackupBin, "--prepare", "--target-dir="+backupDirectory).Output(); err != nil {
```
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go]: muted

File: jobs/mysql_restore_job.go [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+}
+
+func mySQLActive() (bool, error) {
+	// systemctl is-active returns an exit code 0 if service is active, or non-zero otherwise
+	_, err := exec.Command("systemctl", "is-active", "--quiet", mySQLServiceName).Output()
+	if err == nil {
+		return true, nil
+	}
+
+	if _, ok := err.(*exec.ExitError); ok {
```
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go:219]: reworked

File: jobs/mysql_restore_job.go (hunk 5 above)
Comment (oter) [reply jobs/mysql_restore_job.go]: rerr is used :)

File: jobs/mysql_restore_job.go [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+}
+
+func prepareRestoreCommands(
+	ctx context.Context,
+	backupName string,
+	config *BackupLocationConfig,
+	targetDirectory string,
+	stdout io.Writer,
+	stderr io.Writer,
+) (xbcloud, xbstream *exec.Cmd, _ error) {
```
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go:126]: removed

File: jobs/mysql_restore_job.go (hunk 5 above)
Comment (artemgavrilov) [reply jobs/mysql_restore_job.go]: > Can we eliminate named function results?
🙂

Comment (oter) [reply jobs/mysql_restore_job.go]: will `(_, _ *bytes.Buffer, rerr error)` work for you? Since rerr is used in the defer.

Comment (oter) [reply jobs/mysql_restore_job.go]: rerr is left only, came up with leaving as is.

Comment (artemgavrilov) [reply jobs/mysql_restore_job.go]: Ok, seems that we can't eliminate)

File: jobs/mysql_restore_job.go [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+	}
+
+	if errors.As(err, &exec.ExitError{}) {
+		return false, nil
+	}
+
+	return false, err
+}
+
+func waitForMySQL(expectedActiveStatus bool) error {
```
Comment (artemgavrilov) [inline jobs/mysql_restore_job.go:217]: Do we need this? I thought `systemctl start/stop` are blocking operations

File: jobs/mysql_restore_job.go [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+	return filepath.Walk(path, func(name string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		return errors.WithStack(os.Chown(name, uid, gid))
+	})
+}
+
+func mySQLUserAndGroupIDs() (uid, gid int, rerr error) {
```
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go]: eliminated

File: jobs/mysql_restore_job.go [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+		return errors.WithStack(err)
+	}
+	if err := chownRecursive(mySQLDirectory, uid, gid); err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+func (j *MySQLRestoreJob) Run(ctx context.Context, send Send) (rerr error) {
```
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go:325]: comment added

File: jobs/mysql_restore_job.go [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+	}
+
+	if exists, err := isPathExists(mySQLDirectory); err != nil {
+		return errors.WithStack(err)
+	} else if exists {
+		postfix := ".old" + strconv.FormatInt(time.Now().Unix(), 10)
+		if err := os.Rename(mySQLDirectory, mySQLDirectory+postfix); err != nil {
+			return errors.WithStack(err)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go]: done

File: jobs/mysql_restore_job.go (hunk 13 above)
Comment (oter) [reply jobs/mysql_restore_job.go:199]: removed

File: jobs/mysql_restore_job.go [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+		if err == nil {
+			return
+		}
+
+		if rerr != nil {
+			rerr = errors.Wrapf(rerr, "removing temporary directory error: %s", err)
+		} else {
+			rerr = errors.WithStack(err)
+		}
+	}()
```
[the remark this answers is not in the document]
Comment (oter) [reply jobs/mysql_restore_job.go:353]: fixed

File: jobs/mysql_restore_job.go [hunk 18]
```diff
@@ -0,0 +1,397 @@
...
+		if err := stopMySQL(ctx); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	if err := restoreBackup(ctx, tmpDir, mySQLDirectory); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := startMySQL(ctx); err != nil {
```
Comment (BupycHuk) [inline jobs/mysql_restore_job.go:384]: should we start MySQL if it wasn't active?

File: jobs/mysql_restore_job.go [hunk 19]
```diff
@@ -0,0 +1,397 @@
...
+	}
+
+	return nil
+}
+
+func mySQLActive(ctx context.Context) (bool, error) {
+	ctx, cancel := context.WithTimeout(ctx, systemctlTimeout)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, "systemctl", "is-active", "--quiet", mySQLServiceName)
```
Comment (BupycHuk) [inline jobs/mysql_restore_job.go:201]: have you discussed cases where systemd is not installed?

Comment (oter) [reply jobs/mysql_restore_job.go:201]: we are expecting that MySQL runs as a service using systemd.

Comment (BupycHuk) [reply jobs/mysql_restore_job.go:201]: Good, don't forget to mention that in docs

File: jobs/mysql_restore_job.go (hunk 18 above)
Comment (oter) [reply jobs/mysql_restore_job.go:384]: good question, need to ask products about that

**Review by artemgavrilov — APPROVED** [review]

File: jobs/mysql_restore_job.go (hunk 18 above)
Comment (oter) [reply jobs/mysql_restore_job.go:384]: From a technical standpoint, I need to start MySQL to ensure that the restored db files don't break MySQL - e.g., when I try to restore MySQL backup on top of the Percona Distribution for MySQL the last will fail to start.

**Review by BupycHuk — APPROVED** [review]

**Pull request merged**